src/server/debug/server/database_dump.go

```diff
@@ -0,0 +1,139 @@
+			FROM pg_stat_user_tables ORDER BY schemaname, relname;
+		`,
+	}
+	var errs []error
+	for filename, query := range queries {
+		if err := collectDebugFile(tw, filename, "json", func(w io.Writer) error {
+			rows, err := s.database.QueryContext(ctx, query)
+			if err != nil {
+				return errors.EnsureStack(err)
+			}
+			return s.writeRowsToJSON(rows, w)
+		}, prefix...); err != nil {
+			errs = append(errs, errors.Wrapf(err, "failed to query %q", filename))
+		}
+	}
+	if len(errs) > 0 {
+		return fmt.Errorf("%v", errs)
+	}
+	return nil
+}
+
+func (s *debugServer) collectDatabaseTables(ctx context.Context, tw *tar.Writer, prefix ...string) error {
+	tables, err := pachsql.ListTables(ctx, s.database)
+	if err != nil {
+		return errors.EnsureStack(err)
+	}
+	for _, table := range tables {
+		fullPrefix := strings.Join(append(append([]string{}, prefix...), "tables", table.SchemaName), "/")
+		if err := collectDebugFile(tw, table.TableName, "json", func(w io.Writer) error {
+			bufferedWriter := bufio.NewWriter(w)
+			if err := s.collectTable(ctx, bufferedWriter, &table); err != nil {
+				return errors.Wrapf(err, "error collecting table %s.%s", table.SchemaName, table.TableName)
+			}
+			if err := bufferedWriter.Flush(); err != nil {
+				return errors.Wrap(err, "failed to flush buffer")
+			}
+			return nil
+		}, fullPrefix); err != nil {
+			return errors.EnsureStack(err)
+		}
+	}
+	return nil
+}
+
+func (s *debugServer) collectTable(ctx context.Context, w io.Writer, table *pachsql.SchemaTable) error {
+	column, err := pachsql.KeyColumn(ctx, s.database, table)
+	if err != nil {
+		return errors.Wrap(err, "error getting key column")
+	}
+	query := fmt.Sprintf(`SELECT * FROM %s.%s ORDER BY %s ASC`, table.SchemaName, table.TableName, column)
+	rows, err := s.database.QueryContext(ctx, query)
+	if err != nil {
+		return errors.Wrap(err, "error querying")
+	}
+	defer rows.Close()
+	return s.writeRowsToJSON(rows, w)
+}
+
+func (s *debugServer) writeRowsToJSON(rows *sql.Rows, w io.Writer) error {
+	if _, err := w.Write([]byte("[\n\t")); err != nil {
+		return errors.Wrap(err, "could not write table")
+	}
+	wroteFirstRow := false
+	for rows.Next() {
+		if wroteFirstRow {
+			if _, err := w.Write([]byte(",\n\t")); err != nil {
+				return errors.Wrap(err, "could not write table")
+			}
+		}
+		row := pachsql.RowMap{}
+		if err := sqlx.MapScan(rows, row); err != nil {
+			return errors.Wrap(err, "could not map scan rows")
+		}
+		jsonRow, err := json.Marshal(row)
+		if err != nil {
+			return errors.Wrap(err, "could not marshal row")
+		}
+		if _, err := w.Write(jsonRow); err != nil {
+			return errors.Wrap(err, "could not write table")
+		}
+		wroteFirstRow = true
+	}
+	if err := rows.Err(); err != nil {
+		if errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled) {
+			w.Write([]byte(",\n\t{\"message\": \"result set was truncated.\"}\n]"))
+			return nil
+		}
+		return errors.Wrap(err, "error iterating through rows")
```

Comment:
This is an error *after* iterating through rows.  This would be a read error that makes rows.Next falsely determine that there are no more rows, I guess.